### PR TITLE
Lda

### PR DESCRIPTION
**Pull request by mostafa-mahmoud**

### CODE REVIEW

File: lib/LDA.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class LDARecommender(ContentBased):
+    def __init__(self, abstracts, n_factors, n_iterations=5):
+        self.n_factors = n_factors
+        self.n_items = len(abstracts)
+        self.abstracts = abstracts
+        self.n_iterations = n_iterations
+
+    def train(self):
+        tf_vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english', max_features=self.n_items)
+        tf = tf_vectorizer.fit_transform(self.abstracts)
```
Comment (ndomar) [inline lib/LDA.py]: Can we use more intuitive variable name?

File: lib/LDA.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+        lda = LatentDirichletAllocation(n_topics=self.n_factors, max_iter=self.n_iterations,
+                                        learning_method='online', learning_offset=50., random_state=0)
+        self.word_distribution = dist = lda.fit_transform(tf)
```
Comment (ndomar) [inline lib/LDA.py]: I thought the LDA takes the abstracts as an input? The whole term frequency calculation seems redundant to me. We already have these data in MySQL

File: lib/recommender_system.py [hunk 3]
```diff
@@ -25,8 +26,9 @@ def __init__(self):
         self.config = RecommenderConfiguration()
         self.n_factors = self.config.get_hyperparameters()['n_factors']
         self.n_iterations = self.config.get_options()['n_iterations']
+        # self.content_based = ContentBased(self.abstracts, self.n_factors, self.n_iterations)
```
Comment (ndomar) [inline lib/recommender_system.py]: I am not okay with merging commented code into master.

File: lib/LDA.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        # tf_feature_names = tf_vectorizer.get_feature_names()
+        # print([['%.3f:%s' % (lda.components_[topic_idx][i], tf_feature_names[i]) for i in topic.argsort()]
+        #                                        for topic_idx, topic in enumerate(dist)])
```
Comment (ndomar) [inline lib/LDA.py]: I am not okay with merging commented code into master.

File: lib/LDA.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
+import numpy
+from lib.content_based import ContentBased
+from sklearn.decomposition import LatentDirichletAllocation
+from sklearn.feature_extraction.text import CountVectorizer
+
+
+class LDARecommender(ContentBased):
```
Comment (ndomar) [inline lib/LDA.py]: I don't see why this class is called LDARecommender, if it does provide recommendation then it should extend abstract_recommender as well. if it does not then simply rename it to LDA.

File: tests/tests.py [hunk 6]
```diff
@@ -19,20 +18,23 @@ def setUp(self):
         articles_cnt = 4
         users_cnt = 10
 
-        def mock_process(self):
+        def mock_process(self=None):
```
Comment (ndomar) [inline tests/tests.py]: Why is this necessary?

Comment (mostafa-mahmoud) [inline tests/tests.py]: In python 2, it's needed because:
DataParser.process() will correspond to mock_process(DataParser)

In Python 3, it's not needed because:
DataParser.process() will correspond to mock_process() (that's why I'm using a default value)

Comment (mostafa-mahmoud) [inline tests/tests.py]: In the future, we should remove 'self' totally. but I'm still depending on Python 2 sometimes.

File: lib/LDA.py (hunk 5 above)
Comment (mostafa-mahmoud) [inline lib/LDA.py]: Still I didn't merge your AbstractRecommender, I opened issue #8 for it, and will do it later.

File: lib/LDA.py (hunk 2 above)
Comment (mostafa-mahmoud) [inline lib/LDA.py]: The CountVectorizer still has more options that can be used while computing the term frequencies, like skipping some English words and so on, so probably it's better to use theirs since it can fit more potential future changes.

File: lib/LDA.py (hunk 4 above)
Comment (mostafa-mahmoud) [inline lib/LDA.py]: updated.

File: lib/recommender_system.py (hunk 3 above)
Comment (mostafa-mahmoud) [inline lib/recommender_system.py]: updated. It's a default algorithm until others are instantiated.

File: lib/LDA.py [hunk 7]
```diff
@@ -13,12 +13,12 @@ def __init__(self, abstracts, n_factors, n_iterations=5):
...
-        tf_vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english', max_features=self.n_items)
-        tf = tf_vectorizer.fit_transform(self.abstracts)
+        term_freq_vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english', max_features=self.n_items)
+        term_freq = term_freq_vectorizer.fit_transform(self.abstracts)
 
         lda = LatentDirichletAllocation(n_topics=self.n_factors, max_iter=self.n_iterations,
                                         learning_method='online', learning_offset=50., random_state=0)
-        self.word_distribution = dist = lda.fit_transform(tf)
+        self.word_distribution = lda.fit_transform(term_freq)
         # tf_feature_names = tf_vectorizer.get_feature_names()
```
Comment (ndomar) [inline lib/LDA.py]: Can we remove those comments as well?

File: lib/LDA.py [hunk 8]
```diff
@@ -13,12 +13,12 @@ def __init__(self, abstracts, n_factors, n_iterations=5):
         self.n_iterations = n_iterations
 
     def train(self):
-        tf_vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english', max_features=self.n_items)
-        tf = tf_vectorizer.fit_transform(self.abstracts)
+        term_freq_vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english', max_features=self.n_items)
```
Comment (ndomar) [inline lib/LDA.py]: Regardless of the features the CountVectorizer provides, I am pretty sure the vocabulary we have was pre-processed by removing stop words, applying stemming, etc. This is your design and I will merge it but I kinda feel that we are re-inventing the wheel here.

File: lib/LDA.py (hunk 7 above)
Comment (mostafa-mahmoud) [inline lib/LDA.py]: Done.

File: lib/LDA.py (hunk 8 above)
Comment (mostafa-mahmoud) [inline lib/LDA.py]: Okay, we can merge it for now, and later on I will try to check if I can replace it with stuff from the DataParser. opened for this #9
